scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3SchemaBuilder.java

```diff
@@ -12,125 +13,477 @@
-        this.baseTableName = tableName.substring(0, tableName.length() - SCYLLA_CDC_LOG_SUFFIX.length());
-        return this;
-    }
+        String cdcLogTableName = row.getColumnDefinitions().getTable(0);
+        Preconditions.checkArgument(cdcLogTableName.endsWith(SCYLLA_CDC_LOG_SUFFIX));
 
-    public Driver3SchemaBuilder withClusterMetadata(Metadata metadata) {
-        this.metadata = Preconditions.checkNotNull(metadata);
+        this.baseTableName = cdcLogTableName.substring(0, cdcLogTableName.length() - SCYLLA_CDC_LOG_SUFFIX.length());
+        this.cdcColumnDefinitions = row.getColumnDefinitions();
         return this;
     }
 
-    public ChangeSchema build() {
-        generatePrimaryKeyColumns();
-
-        return new ChangeSchema(generateChangeSchemaColumns());
-    }
-
-    private void generatePrimaryKeyColumns() {
+    public Driver3SchemaBuilder withClusterMetadata(Metadata metadata) throws UnresolvableSchemaInconsistencyException {
         Preconditions.checkNotNull(keyspace);
         Preconditions.checkNotNull(baseTableName);
         Preconditions.checkNotNull(metadata);
 
-        /*
-         * This is a potentially really dangerous operation. Rationale:
-         * 1) Metadata (returned by session.getCluster().getMetadata()) might be stale.
-         * 2) Even if Metadata was always up-to-date, we could have a race condition:
-         *    a) session.getCluster().getMetadata()
-         *    b) Do a query on CDC log (which returns other metadata)
-         * If the metadata changed between call a) and b) there could be a discrepancy
-         * between them. However, in this place we are only reading what
-         * columns are in primary key, which cannot be changed. But, there could
-         * be a situation when someone drops the table and quickly recreates it (with
-         * different schema) and this operation took place between call a) and b).
-         * In such case, this code is not correct!
-         *
-         * Ideally, the server should send the corresponding base table metadata together with the CDC log table metadata,
-         * and we'll be able to drop this hack. See Scylla issue #7824.
-         */
-        TableMetadata baseTableMetadata = baseTableMetadata();
-        baseTablePartitionKeyColumnNames = baseTableMetadata.getPartitionKey().stream().map(ColumnMetadata::getName).collect(Collectors.toSet());
-        baseTableClusteringKeyColumnNames = baseTableMetadata.getClusteringColumns().stream().map(ColumnMetadata::getName).collect(Collectors.toSet());
+        KeyspaceMetadata keyspaceMetadata = metadata.getKeyspace(keyspace);
+        if (keyspaceMetadata == null) {
+            throw new UnresolvableSchemaInconsistencyException(
+                    String.format("Keyspace '%s' is missing from the cluster metadata.", keyspace));
+        }
+
+        TableMetadata tableMetadata = keyspaceMetadata.getTable(baseTableName);
```

Comment:
I think it's not desirable that withRow has to be called before withClusterMetadata can be called. Otherwise baseTableName will be null here. Usually builders allow to call all the methods in random order. If withRow is required then let's take row in a constructor of the builder to make sure that baseTableName is always set when other methods are called.

Reply:
In the amended commit changed it to a factory. There is now only a single static `getChangeSchema` method.